[SQOOP-2906] Optimize toAvroIdentifier

JoeriHermans opened the pull request:
Hi all

Our distributed profiler indicated some inefficiencies in the AvroUtil.toAvroIdentifier method, more specifically, the use of Regex patterns. This can be directly observed from the FlameGraph generated by this profiler (https://jhermans.web.cern.ch/jhermans/sqoop_avro_flamegraph.svg). We implemented an optimization, and compared this with the original method. On our testing machine, the optimization by itself is 500% (on average) more efficient compared to the original implementation. We have yet to test how this optimization will influence the performance of user jobs.

Any suggestions or remarks are welcome.

Kind regards,

Joeri

Special thanks to @dlanza1 for the second optimization! :+1: 

**[edit]**
We tested a custom JAR with this method on our cluster, and we have an increased throughput (writing to Parquet) of 200-300% depending on the load.

Commit:
Optimize toAvroIdentifier

Our distributed profiler indicated some inefficiencies in the AvroUtil.toAvroIdentifier method, more specifically, the use of Regex patterns. This can be directly observed from the FlameGraph generated by this profiler (https://jhermans.web.cern.ch/jhermans/sqoop_avro_flamegraph.svg). We implemented an optimization, and compared this with the original method. On our testing machine, the optimization by itself is 230% (on average) more efficient compared to the original implementation. We have yet to test how this optimization will influence the performance of user jobs.

Commit:
Optimize toAvroIdentifier (2)

Optimization done by @dlanza1. This optimization is even more efficient (and more readable) than the previous one.

Commit:
Update code-style AvroUtil.toAvroIdentifier

stanleyxu2005 commented on src/java/org/apache/sqoop/avro/AvroUtil.java:
```diff
@@ -114,11 +114,20 @@ public static String toAvroColumn(String column) {
-    String formattedCandidate = candidate.replaceAll("\\W+", "_");
-    if (formattedCandidate.substring(0,1).matches("[a-zA-Z_]")) {
-      return formattedCandidate;
+    char[] data = candidate.toCharArray();
```
I'd suggest you create a private function `replaceWhitespace` and move your code logic there. So it will be

```
  String formattedCandidate = replaceWhitespace(candidate);
```

Second how can you make sure the string data has a valid length, so that `data[0]` is valid?

JoeriHermans commented:
@stanleyxu2005 Why would you check if data has valid length? The original code also crashes if an empty string is specified, so I didn't bother with checking for validity. Furthermore, I don't think it is necessary to create a wrapper for this method.

I also wanted to ask why in https://github.com/apache/sqoop/commit/1dd50cfb2ae327b0df8393dd96d1adb86bb2f65f candidate.replaceAll("\W+", "") has been replaced with candidate.replaceAll("\W+", "_")? This change makes our code a bit slower, and does not add any value from my point of view. Of course, if this is done with a particular reason, I will commit the change.

[edit] Ignore the comment made about https://github.com/apache/sqoop/commit/1dd50cfb2ae327b0df8393dd96d1adb86bb2f65f, will commit new version.

Kind regards,

Joeri

stanleyxu2005 commented on src/java/org/apache/sqoop/avro/AvroUtil.java:
```diff
@@ -114,11 +114,20 @@ public static String toAvroColumn(String column) {
+
+    char initial = data[0];
+    if (Character.isLetter(initial) || initial == '_') {
+      return new String(data, 0, stringIndex);
```
Your code will first create a char array and then eventually update char in the array. As result you will create another copy as a new String. Have you thought about using a `StringBuilder` directly?

```
  final StringBuilder sb = new StringBuilder();
  for (char c : candidate) {
    if (Character.isLetterOrDigit(c) || c == '_') {
      sb.append(c);
    }
  }
  ...
    return sb.toString();
```

JoeriHermans commented:
@stanleyxu2005 The main difference in my implementation is that I only have to do a single copy. StringBuilders are inefficient, char arrays are a lot faster since everything is pre-allocated already. Furthermore, the overhead over constantly allocating a new object, and the fact that internally the StringBuilder will do some copying as well, make this an inefficient approach.

For example, I implemented your approach and instead of a 500% improvement with our method, I have a 230% increase in performance with the StringBuilder approach. So the char-array is definitely a lot faster, and this is what it is all about, since this is a very prominent function during the conversion to Avro.

Kind regards,

Joeri

Commit:
Update AvroUtil.java

Comply to https://github.com/apache/sqoop/commit/1dd50cfb2ae327b0df8393dd96d1adb86bb2f65f regex.

szvasas commented:
Closing this PR since the corresponding JIRA seems to be closed already.
